contracts/Strategy.sol

```diff
@@ -53,80 +234,648 @@ contract Strategy is BaseStrategy {
+        if (_amountNeeded <= wantBalance) {
+            return (_amountNeeded, 0);
+        }
+
+        uint256 amountToFree = _amountNeeded.sub(wantBalance);
+        uint256 deposited = estimatedTotalAssets().sub(wantBalance);
+
+        if (amountToFree > deposited) {
+            amountToFree = deposited;
+        }
+
+        if (amountToFree > 0) {
+            uint256 sharesNeeded = wantToBentoShares(amountToFree);
+            uint256 bentoShares = sharesInBento();
+
+            if (sharesNeeded > bentoShares) {
+                uint256 sharesToFreeFromKashi = sharesNeeded.sub(bentoShares);
+                uint256 sharesFreedFromKashi = 0;
+
+                // Find the lowest apr pair with at least the lesser of
+                //   - the amount to free
+                //   - the mean assets per pair
+                (IKashiPair lowestPair, uint256 lowestPid) =
+                    lowestInterestPair(
+                        Math.min(
+                            sharesToFreeFromKashi,
+                            wantToBentoShares(
+                                estimatedTotalAssets().div(kashiPairs.length)
+                            )
+                        )
+                    );
+                if (address(lowestPair) != address(0)) {
+                    sharesFreedFromKashi = liquidateKashiPair(
+                        lowestPair,
+                        lowestPid,
+                        sharesToFreeFromKashi
+                    );
+                }
+
+                for (
+                    uint256 i = 0;
+                    i < kashiPairs.length &&
+                        sharesToFreeFromKashi > sharesFreedFromKashi;
+                    i++
+                ) {
+                    KashiPairInfo memory kashiPairInfo = kashiPairs[i];
+
+                    if (address(kashiPairInfo.kashiPair) == address(lowestPair))
+                        continue; // we already visited this
+
+                    sharesFreedFromKashi = sharesFreedFromKashi.add(
+                        liquidateKashiPair(
+                            kashiPairInfo.kashiPair,
+                            kashiPairInfo.pid,
+                            sharesToFreeFromKashi.sub(sharesFreedFromKashi)
+                        )
+                    );
+                }
+            }
+
+            bentoBox.withdraw(
+                BIERC20(address(want)),
+                address(this),
+                address(this),
+                0,
+                sharesInBento()
+            );
+        }
+
+        _liquidatedAmount = Math.min(balanceOfWant(), _amountNeeded);
+
+        if (_amountNeeded > _liquidatedAmount) {
+            _loss = _amountNeeded.sub(_liquidatedAmount);
+        }
+    }
+
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _liquidatedAmount)
+    {
+        (_liquidatedAmount, ) = liquidatePosition(
+            estimatedTotalAssets().sub(balanceOfWant())
```

Comment:
Should this be actually only estimatedTotalAssets() since the spec on this claims this:

```
  /**
     * Liquidate everything and returns the amount that got freed.
     * This function is used during emergency exit instead of `prepareReturn()` to
     * liquidate all of the Strategy's positions back to the Vault.
     */
     ```
     
    and on loc 313 we are substracting again balanceOfWant from the amount, so would that lead into accounting errors?

Reply:
Right